# [자연어처리]
# 14주차(14-2). GPT

# 202002961 김현주
* **모든 셀 실행** 후 제출하시기 바랍니다.
* **실습 (14-2-1)**이 있습니다.

> ### 중간 중간 **'### ... 이 부분을 완성하시오'** 라는 부분의 코드를 완성해 보기 바랍니다. (강의 자료를 참고하여 '???' 부분을 작성해 보시기 바랍니다.)

참고 자료 : https://huggingface.co/blog/how-to-generate

사전 학습된 한국어 GPT-2를 이용하여 다음 문장 예측을 실습해봅시다.

In [ ]:
! pip install transformers ### ... 이 부분을 완성하시오

# [런타임] - [런타임 유형 변경] - [T4 GPU]

# 1. KoGPT2로 문장 생성하기

transformers 패키지를 사용하여 모델과 토크나이저를 로드합니다. BERT와 마찬가지로 GPT는 이미 누군가가 학습해둔 모델을 사용하는 것이므로 우리가 사용하는 모델과 토크나이저는 항상 맵핑 관계여야 합니다.

In [ ]:
import tensorflow as tf
from transformers import AutoTokenizer
from transformers import TFGPT2LMHeadModel

TFGPT2LMHeadModel.from_pretrained('GPT 모델 이름')을 넣으면 두 개의 문장이 이어지는 문장 관계인지 여부를 판단하는 GPT 구조를 로드합니다. AutoTokenizer.from_pretrained('모델 이름')을 넣으면 해당 모델이 학습되었을 당시에 사용되었던 토크나이저를 로드합니다.

In [ ]:
tokenizer = AutoTokenizer.from_pretrained('skt/kogpt2-base-v2') ### ... 이 부분을 완성하시오

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [ ]:
model = TFGPT2LMHeadModel.from_pretrained('skt/kogpt2-base-v2', from_pt=True) ### ... 이 부분을 완성하시오

Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFGPT2LMHeadModel: ['lm_head.weight', 'transformer.h.3.attn.masked_bias', 'transformer.h.7.attn.masked_bias', 'transformer.h.2.attn.masked_bias', 'transformer.h.4.attn.masked_bias', 'transformer.h.1.attn.masked_bias', 'transformer.h.0.attn.masked_bias', 'transformer.h.11.attn.masked_bias', 'transformer.h.9.attn.masked_bias', 'transformer.h.6.attn.masked_bias', 'transformer.h.10.attn.masked_bias', 'transformer.h.5.attn.masked_bias', 'transformer.h.8.attn.masked_bias']
- This IS expected if you are initializing TFGPT2LMHeadModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFGPT2LMHeadModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassifica

> 일반적으로 PyTorch 모델을 TensorFlow 모델로 변환하는 과정에서 흔히 발생하는 워닝 메시지. 변환 과정에서 모든 가중치가 완벽하게 일치하지 않을 수 있다는 것을 의미. 대부분의 경우 모델은 제대로 작동.

GPT가 생성할 문장의 방향성을 알려주기 위해서 시작 문자열을 정해줍니다. 여기서는 '건강한 몸을 만들기 위해서는' 이라는 문자열을 주고 GPT에게 이어서 문장을 생성해보라고 해봅시다.

In [ ]:
sent =  '건강한 몸을 만들기 위해서는'  ### ... 이 부분을 완성하시오

In [ ]:
input_ids = tokenizer.tokenize(sent) ### ... 이 부분을 완성하시오
print(input_ids) ### ... 이 부분을 완성하시오

['▁건강한', '▁몸을', '▁만들기', '▁위해서는']


GPT의 입력으로는 정수 인코딩 된 결과가 입력되어야 하므로 tokenizer.encode()를 통해서 '건강한 몸을 만들기 위해서는'이라는 문자열을 정수 시퀀스로 변환해줍니다.

In [ ]:
input_ids = tokenizer.encode(sent) ### ... 이 부분을 완성하시오
input_ids = tf.convert_to_tensor([input_ids])
print(input_ids) ### ... 이 부분을 완성하시오

tf.Tensor([[22159 14226 15381 11357]], shape=(1, 4), dtype=int32)


22159 14226 15381 11357라는 4개의 정수 시퀀스를 얻습니다.

해당 정수 시퀀스를 GPT의 입력으로 사용하여 GPT가 이어서 문장을 생성하도록 해봅시다. 주어진 문장으로부터 이어서 문장을 생성하도록 하는 것은 model.generate()를 사용합니다.

In [ ]:
output = model.generate(input_ids,
                   max_length=128,
                   repetition_penalty=2.0) ### ... 이 부분을 완성하시오
output_ids = output.numpy().tolist()[0]
print(output_ids) ### ... 이 부분을 완성하시오

[22159, 14226, 15381, 11357, 23879, 39306, 25235, 9376, 33820, 10362, 9430, 15967, 38594, 12443, 8137, 17764, 9033, 9199, 15177, 26421, 387, 15386, 409, 7162, 9346, 7071, 11913, 9091, 9377, 12486, 10108, 24873, 7198, 9391, 37611, 9676, 14458, 9492, 20305, 10090, 10254, 9178, 7335, 8704, 41845, 18418, 9277, 18324, 8643, 7122, 8135, 9351, 16709, 10335, 18380, 16286, 13940, 19671, 9048, 18281, 8648, 44869, 6824, 12923, 13600, 9600, 39137, 8066, 23865, 8470, 10461, 9164, 7609, 8397, 9225, 9029, 45837, 9052, 48276, 15528, 10386, 10988, 8711, 22452, 8765, 27929, 12972, 12364, 23511, 8450, 16803, 21302, 6903, 9410, 13041, 25606, 49019, 7586, 9752, 8143, 10850, 10007, 26006, 27147, 8146, 26523, 9359, 9341, 13219, 35943, 14441, 8000, 13532, 13565, 11634, 12268, 10279, 36122, 9960, 8405, 11545, 13235, 7849, 17304, 14355, 10826, 43075, 7235]


기존의 22159 14226 15381 11357 라는 4개의 정수 시퀀스 뒤에도 여러 정수들이 추가로 생성된 것을 볼 수 있습니다. 정수들이 단순히 나열된 것만으로는 GPT가 실제로 어떤 문장을 생성했는지 알기 어려우니 해당 정수 시퀀스를 한국어로 변환해봅시다. 이 과정은 tokenizer.decode()를 사용하여 가능합니다.

In [ ]:
print(tokenizer.convert_ids_to_tokens(output_ids))

['▁건강한', '▁몸을', '▁만들기', '▁위해서는', '▁무엇보다', '▁규칙적인', '▁운동과', '▁식', '이요', '법을', '▁통해', '▁균형', '▁잡힌', '▁영양', '을', '▁섭취', '하는', '▁것이', '▁중요하다.\n', '특히', ',', '▁비타민', 'C', '는', '▁체', '내', '▁에너지', '▁생', '성을', '▁촉진', '시켜', '▁신진', '대', '사를', '▁원활하게', '▁하고', '▁면역', '력을', '▁높여', '주는', '▁역할을', '▁한다.\n', '또', '한', '▁항산화', '▁작용을', '▁하는', '▁폴리', '페', '놀', '은', '▁우리', '▁몸의', '▁세포', '막을', '▁구성하는', '▁필수', '▁성분', '이다.\n', '폴리', '펩', '타이드', '가', '▁풍부한', '▁식품', '으로는', '▁토마토', '와', '▁브로', '콜', '리,', '▁양', '배', '추', '▁등이', '▁있다.\n', '토마토', '에는', '▁식이', '섬유', '소가', '▁풍부', '해', '▁피로', '회', '복에', '▁도움을', '▁준다.\n', '브로', '컬', '리에는', '▁칼슘', '과', '▁철', '분이', '▁풍부하다.\n', '비타', '민', '▁C', '의', '▁흡', '수를', '▁돕는', '▁칼륨', '이', '▁풍부하게', '▁들어', '▁있어', '▁뼈', '▁건강에', '▁좋다.\n', '양', '파는', '▁섬유', '질이', '▁많아', '▁혈', '압을', '▁낮', '춰', '주고', '▁혈액', '순', '환을', '▁도와', '준다.\n', '마늘', '도']


In [ ]:
tokenizer.decode(output_ids)

'건강한 몸을 만들기 위해서는 무엇보다 규칙적인 운동과 식이요법을 통해 균형 잡힌 영양을 섭취하는 것이 중요하다.\n특히, 비타민C는 체내 에너지 생성을 촉진시켜 신진대사를 원활하게 하고 면역력을 높여주는 역할을 한다.\n또한 항산화 작용을 하는 폴리페놀은 우리 몸의 세포막을 구성하는 필수 성분이다.\n폴리펩타이드가 풍부한 식품으로는 토마토와 브로콜리, 양배추 등이 있다.\n토마토에는 식이섬유소가 풍부해 피로회복에 도움을 준다.\n브로컬리에는 칼슘과 철분이 풍부하다.\n비타민 C의 흡수를 돕는 칼륨이 풍부하게 들어 있어 뼈 건강에 좋다.\n양파는 섬유질이 많아 혈압을 낮춰주고 혈액순환을 도와준다.\n마늘도'

'건강한 몸을 만들기 위해서는' 이라는 문자열에 이어서 그 뒤에 건강한 몸을 만들기 위한이라는 문맥에 맞는듯한 글들을 생성합니다.

물론 GPT가 생성한 문장들은 문맥상 그럴듯해 보이지만 실제 사실 여부와는 다를 수 있으니 이 점은 늘 주의해야 합니다.

# 2. Numpy로 Top 5 뽑기

GPT는 기본적으로 이전 단어들로부터 다음 단어를 예측하는 언어 모델(Language Model)입니다. 위의 실습에서 확인한 바와 같이 '건강한 몸을 만들기 위해서는' 이라는 입력을 넣었을 때 GPT는 다음 단어로 '무엇보다' 라는 단어를 예측했었는데요.


실제로는 수많은 후보의 다음 단어들이 있었지만, 그 중 가장 확률이 높은 단어. 즉, Top 1의 단어인 '무엇보다'를 예측한 것입니다. 그렇다면 다음 단어로 또 어떤 후보들이 있었는지 Top 5의 단어들을 뽑아봅시다.


model()에다가 '건강한 몸을 만들기 위해서는'의 정수 인코딩 된 결과를 입력으로 넣은 후 가장 확률이 높은 Top 5의 단어들을 뽑아냅니다.

In [ ]:
import numpy as np

In [ ]:
output = model(input_ids)

In [ ]:
output.logits ### ... 이 부분을 완성하시오

<tf.Tensor: shape=(1, 4, 51200), dtype=float32, numpy=
array([[[-6.159608  , -5.6269383 , -5.299071  , ...,  0.18111259,
         -4.966682  , -3.093627  ],
        [-4.6646886 , -5.924711  , -4.8221297 , ...,  2.4812746 ,
         -3.5151458 ,  2.3900423 ],
        [-5.085371  , -4.0964794 , -6.0712833 , ...,  1.1920143 ,
         -2.9018292 , -3.3687859 ],
        [-6.2043056 , -6.2846212 , -6.3270826 , ...,  1.8252188 ,
         -3.9164739 , -2.716536  ]]], dtype=float32)>

In [ ]:
output.logits.shape

TensorShape([1, 4, 51200])

In [ ]:
top5 = tf.math.top_k(output.logits[0, -1], k=5)
top5

TopKV2(values=<tf.Tensor: shape=(5,), dtype=float32, numpy=
array([10.895332, 10.395889,  9.945644,  9.840966,  9.831181],
      dtype=float32)>, indices=<tf.Tensor: shape=(5,), dtype=int32, numpy=array([23879, 22159, 20030, 11342, 39306], dtype=int32)>)

그 후 Top 5의 단어를 한국어로 변환하여 출력해봅시다.

In [ ]:
tokenizer.convert_ids_to_tokens(top5.indices.numpy()) ### ... 이 부분을 완성하시오

['▁무엇보다', '▁건강한', '▁무엇보다도', '▁건강', '▁규칙적인']

'무엇보다' 라는 단어 외에도 '건강한', '무엇보다도', '건강', '규칙적인' 이라는 4개의 단어가 높은 확률로 선택될 수 있었음을 보여줍니다.

# 3. Numpy Top 5 중 랜덤 단어로 문장 생성해보기

앞서 문장을 생성했을 당시에는 각 시점(time step)마다 가장 확률이 높은 단어를 예측했지만, 이번에는 매 시점마다 Top 5개의 단어들 중에서 랜덤으로 선택하는 방식을 택하여 문장을 생성해봅시다.

In [ ]:
sent = '근육이 커지기 위해서는' ### ... 이 부분을 완성하시오
input_ids = tokenizer.encode(sent) ### ... 이 부분을 완성하시오

In [ ]:
import random

while len(input_ids) < 50:
    output = model(np.array([input_ids]))
    top5 = tf.math.top_k(output.logits[0, -1], k=5)
    token_id = random.choice(top5.indices.numpy())
    input_ids.append(token_id)

In [ ]:
tokenizer.decode(input_ids) ### ... 이 부분을 완성하시오

'근육이 커지기 위해서는 반드시 건강검진과 더불어 규칙적인 운동과 함께 식단조절이 필수이다.\n특히, 평소 식습관이 좋지 않은 경우에는 체지방을 줄이는데 더욱 힘써야만 다이어트에 효과적이다.\n최신형 햄버거는 체지방을 줄이는 데 도움을 줄'

# 4. pipeline을 이용해 문장 생성해보기

In [ ]:
from transformers import pipeline ### ... 이 부분을 완성하시오
pipe = pipeline("text-generation", model="skt/kogpt2-base-v2", device=0) ### ... 이 부분을 완성하시오

In [ ]:
sent = "건강한 몸을 만들기 위해서는"
generated_text = pipe(sent, max_new_tokens=256)  ### ... 이 부분을 완성하시오
print(generated_text[0]['generated_text'])

건강한 몸을 만들기 위해서는 무엇보다도 자신의 몸 상태를 정확히 파악하고 운동하고 꾸준한 운동을 하는 것이 중요하다.
만약 너무 무리한 운동을 시작하게 되면, 관절 주변의 근육에 무리를 주고 관절 운동이 부족해져 무릎 관절에 통증이 생길 수 있다.
관절에 무리를 주는 운동을 할 때, 통증을 예방하기 위해 적절한 스트레칭을 병행해줘야 한다.
관절 주변 근육이 지나치게 긴장하게 되면, 근육을 너무 적게 사용해 관절을 압박하게 된다.
이때 관절 주변 근육을 강화시켜 주지 않도록 스트레칭을 해주면서 관절, 관절을 유연하게 만들어주어 관절 운동이 제대로 이루어지는 것을 확인할 수 있다.
이때 관절을 유연하게 유지 할 수 있도록 관절내시경으로 관절을 자극해주는 운동을 꾸준히 해주는 것이 중요하다.
관절의 근력 강화 운동인 'Selte Clair 운동'은 관절염과 디스크의 치료를 위하여 주로 사용된다.
관절을 자극하는 운동은 운동 전후로 관절을 이완해주는 동작을 함으로써 관절 운동을 도와준다.
또한 간단한 스트레칭으로도 관절 운동의 효과를 볼 수 있어 관절 운동 효과를 극대화 할 수 있다.
강동경희대병원 정형외과 신응수 교수는 "관절운동의 경우 적절한 운동과 적절한 운동시간만 준수하면 관절 주위의 근육과 인대 등을 강화시킬 수 있어 많은 도움을 받는다"고 말했다. 기상청은 이날 전국이 대체로 맑겠다고 밝혔다.
하지만 강원 산간지역을 중심으로 아침 안개특보가 내려


In [ ]:
sent = "안녕하세요"
generated_text = pipe(sent, max_new_tokens=256)
print(generated_text[0]['generated_text'])

안녕하세요?'
소녀는 문득 고개를 끄덕였다.
그녀는 문득 자신이 궁금해 하더니 그제서야 이렇게 물었다.
"이래서는 안 되겠죠?"
소녀가 고개를 끄덕였다.
"우선 내일을 위해 미리 준비하도록 하죠. 오늘 오후 7시(토요일)에는 저한테 연락을 드리고 있소. 지금 그분은 전화를 받지 않고 있소."
소녀가 이렇게 말하자 소여는 잠시 고민하지 않을 수 없었다.
문득 소녀는 지금이 바로 자신의 마음을 드러내려는 적기란 것을 눈치 채고 있었다.
갑자기 소여는 눈앞에서 일어난 일 때문에 괜히 걱정이 돼서 그녀를 올려다보며 말했다.
"이런, 무슨 일이야?"
그러나 소여는 아무 말도 하지 않았다.
소여는 그녀가 대답을 기다리자 아무 말 없이 그녀를 바라보았다.
소여도 그녀를 똑바로 바라보았다.
"어느 누구도 그런 얘기를 하지 않는데, 이젠 내가 어떤 일을 할 수 있을지도 모르는 일이고, 게다가 지금까지는 내가 아무것도 할 줄 몰랐기 때문에 정말 어림도 없는 일이지요? 그런데 이게 무슨 일이에요?"
소여는 그녀를 흘겨보며 다시 물었다.
"정말 말이 잘 되지 아니했는데요."
"어떻게 하면 좋을지 생각해 보게."
"그동안 내가 아무 말도 못하게 방해한 것은 우리에 대한 것이었


In [ ]:
from transformers import pipeline

pipe = pipeline("text-generation", model="heegyu/polyglot-ko-1.3b-chat", device=0) ### ... 이 부분을 완성하시오

In [ ]:
chat_sent = '건강한 몸을 만들기 위해서는'
generated_text = pipe(f"### 사용자: {chat_sent}\n\n### 챗봇:", max_new_tokens=256, eos_token_id=2, pad_token_id=2) ### ... 이 부분을 완성하시오
print(generated_text[0]['generated_text'])

### 사용자: 건강한 몸을 만들기 위해서는

### 챗봇: 건강한 몸을 만들기 위해서는 다음과 같은 방법을 시도해 볼 수 있습니다.

1. 규칙적인 운동: 규칙적인 운동은 신체의 대사를 촉진하고, 근육을 강화하며, 체중을 조절하는 데 도움을 줍니다.

2. 건강한 식단: 건강한 식단은 영양소를 균형 있게 섭취할 수 있도록 도와줍니다. 과일, 채소, 단백질, 지방 등을 골고루 섭취하는 것이 중요합니다.

3. 충분한 수면: 충분한 수면을 취하는 것은 건강한 몸을 유지하는 데 도움이 됩니다. 수면 부족은 스트레스와 피로를 유발할 수 있으므로 충분한 수면을 취하는 것이 중요합니다.

4. 스트레스 관리: 스트레스는 신체의 기능을 저하시키고, 면역체계를 약화시킬 수 있습니다. 스트레스를 관리하는 방법을 찾아서 스트레스를 줄이는 것이 중요합니다.

5. 충분한 휴식: 충분한 휴식은 신체의 기능을 회복시키고, 스트레스를 완화시키는 데 도움이 됩니다. 충분한 휴식을 취하는 것이 중요합니다.

6. 적절한 운동: 적절한 운동은 신체의 대사를 촉진하고,


In [ ]:
chat_sent = '건강한 몸을 만들기 위해서는 어떻게 해야 돼?' ### ... 이 부분을 완성하시오
generated_text = pipe(f"### 사용자: {chat_sent}\n\n### 챗봇:", max_new_tokens=256, eos_token_id=2, pad_token_id=2)
print(generated_text[0]['generated_text'])

### 사용자: 건강한 몸을 만들기 위해서는 어떻게 해야 돼?

### 챗봇: 건강한 몸을 만들기 위해서는 다음과 같은 방법을 시도해 볼 수 있습니다.

1. 규칙적인 운동: 규칙적인 운동은 건강한 몸을 유지하는 데 도움이 됩니다. 하루에 30분 이상의 유산소 운동을 규칙적으로 하는 것이 좋습니다.

2. 충분한 수면: 충분한 수면을 취하는 것은 건강한 몸을 유지하는 데 도움이 됩니다. 수면 부족은 스트레스와 피로를 유발할 수 있으므로 충분한 수면을 취하는 것이 중요합니다.

3. 균형 잡힌 식단: 건강한 식단은 건강한 몸을 유지하는 데 도움이 됩니다. 균형 잡힌 식단은 영양소를 충분히 섭취할 수 있도록 도와주며, 영양소의 불균형을 예방할 수 있습니다.

4. 충분한 휴식: 충분한 휴식은 건강한 몸을 유지하는 데 도움이 됩니다. 충분한 휴식은 스트레스를 줄이고, 면역력을 향상시키는 데 도움이 됩니다.

5. 스트레스 관리: 스트레스는 건강에 해로운 영향을 미칠 수 있으므로, 스트레스를 관리하는 것이 중요합니다. 스트레스를 관리하는 방법으로는 명상, 운동, 취미활동 등


In [ ]:
chat_sent = '건강한 몸을 만들기 위해서는 어떻게 해야 돼?'
prompt = '당신은 AI 챗봇입니다. 사용자에게 도움이 되고 유익한 내용을 제공해야 합니다. 답변은 길고 자세하며 친절한 설명을 덧붙여서 작성하세요.'
generated_text = pipe(f"{prompt}\n\n### 사용자: {chat_sent}\n\n### 챗봇:", max_new_tokens=256, eos_token_id=2, pad_token_id=2)
print(generated_text[0]['generated_text'])

당신은 AI 챗봇입니다. 사용자에게 도움이 되고 유익한 내용을 제공해야 합니다. 답변은 길고 자세하며 친절한 설명을 덧붙여서 작성하세요.

### 사용자: 건강한 몸을 만들기 위해서는 어떻게 해야 돼?

### 챗봇: 건강한 몸을 만들기 위해서는 다음과 같은 방법을 고려해 볼 수 있습니다.

1. 규칙적인 운동: 규칙적인 운동은 신체의 건강을 유지하는 데 도움이 됩니다. 하루에 30분 이상의 유산소 운동을 포함하여, 근력 운동과 유산소 운동을 함께 진행하는 것이 좋습니다.

2. 충분한 수면: 충분한 수면을 취하는 것은 건강한 몸을 유지하는 데 도움이 됩니다. 수면 부족은 신체의 면역 기능을 저하시키고, 질병에 대한 저항력을 약화시킬 수 있습니다.

3. 균형 잡힌 식단: 균형 잡힌 식단은 건강한 몸을 유지하는 데 도움이 됩니다. 과일, 채소, 단백질, 지방 등 다양한 영양소를 균형 있게 섭취하는 것이 중요합니다.

4. 스트레스 관리: 스트레스는 신체의 건강을 유지하는 데 방해가 될 수 있습니다. 스트레스를 관리하는 방법으로는 명상, 운동, 취미 활동 등이 있습니다.

5. 충분한 휴식: 충분한 휴식은 신체의 건강을 유지하는 데 도움이 됩니다. 충분한 수면, 운동, 휴식 등을 통해 스트레스를 관리하고


In [ ]:
print(generated_text[0]['generated_text'].split('### 챗봇:')[1])

 건강한 몸을 만들기 위해서는 다음과 같은 방법을 고려해 볼 수 있습니다.

1. 규칙적인 운동: 규칙적인 운동은 신체의 건강을 유지하는 데 도움이 됩니다. 하루에 30분 이상의 유산소 운동을 포함하여, 근력 운동과 유산소 운동을 함께 진행하는 것이 좋습니다.

2. 충분한 수면: 충분한 수면을 취하는 것은 건강한 몸을 유지하는 데 도움이 됩니다. 수면 부족은 신체의 면역 기능을 저하시키고, 질병에 대한 저항력을 약화시킬 수 있습니다.

3. 균형 잡힌 식단: 균형 잡힌 식단은 건강한 몸을 유지하는 데 도움이 됩니다. 과일, 채소, 단백질, 지방 등 다양한 영양소를 균형 있게 섭취하는 것이 중요합니다.

4. 스트레스 관리: 스트레스는 신체의 건강을 유지하는 데 방해가 될 수 있습니다. 스트레스를 관리하는 방법으로는 명상, 운동, 취미 활동 등이 있습니다.

5. 충분한 휴식: 충분한 휴식은 신체의 건강을 유지하는 데 도움이 됩니다. 충분한 수면, 운동, 휴식 등을 통해 스트레스를 관리하고


## **[[실습 14-2-1]]** 단순 **자연어 생성**과 **챗봇**을 위한 GPT 모델 비교하기

**자연어 생성**을 목적으로 하는 GPT 모델은 다음과 같다.
- model=**"skt/kogpt2-base-v2"**
```
pipe = pipeline("text-generation", model="skt/kogpt2-base-v2", device=0)
```

**챗봇** 용도의 GPT 모델은 다음과 같다.
- model=**"heegyu/polyglot-ko-1.3b-chat"**
```
pipe = pipeline("text-generation", model="heegyu/polyglot-ko-1.3b-chat", device=0)
```
<br>

동일한 문장을 각 pipe의 입력으로 사용해보고, 두 결과를 비교해 보시오.



---

[테스트 입력 문장1]
- (**자연어 생성**에 적합할 것으로 기대되는) 입력 문장1 : "자연어 처리에서 최근"

- **자연어 생성 (sent)** 출력 결과:

    자연어 처리에서 최근 각광받고 있는 ‘디지틀 프로그래밍 언어’라며 컴퓨터 언어의 중요성을 강조했다.
    이에 따라 이 책은 컴퓨터를 제대로 활용할 수 있는 컴퓨터 교육 방안을 제시하고 있다.
    또 컴퓨터와 컴퓨터를 통해 학습할 수 없는 다양한 문제에 대한 효과적인 해결책을 제시해 컴퓨터 교육을 효과적으로 도와준다.
    특히 지난해 출간된 ‘컴퓨터와 세상을 읽는 방법’ 시리즈와 2004년 초판 등에 이어 이번 학기에는 컴퓨터 교육과 관련된 서적들을 새롭게 포함시킨 ‘컴퓨터 세상을 읽는 방법’편을 새롭게 선보이는 등 전자책을 통한 독서의 중요성을 강조하고 있다.
    책자에는 ‘디지틀 프로그래밍 언어’에 대한 내용뿐 아니라 일상생활에서 접하는 컴퓨터 교육과 관련된 다양한 정보를 싣는다.
    이 밖에도 컴퓨터 교육을 올바르게 활용할 수 있도록 초등학교 5학년과 고등학교 2학년 과정을 위한 컴퓨터 교육과정과, 컴퓨터 교육을 위해 노력하는 사람들을 위한 컴퓨터 교육, 생활 속에서 자연스럽게 배울 수 있는 컴퓨터 사용법 소개, 컴퓨터 교육 방법 등이 자세하게 소개돼 있다. 이달부터 한국항공우주산업이 발사 후 2주 만에 궤도 진입에 성공하면서 본격적인 우주개발에 박차를 가한다.
    한국항공우주산업(KAI)이 한국형 발사체인 ‘KA-3’ 제작 사업에 나선다.
    한국항공우주산업은 13일 서울 코엑스 인터컨티넨탈 호텔에서 기자 간담회를 통해 위성체 개발 등 우주개발 사업을 본격화한다고 밝혔다.
    이번 사업으로 KAI

- **챗봇 (chat_sent)** 출력 결과:

    자연어 처리에서 최근에는 텍스트 데이터를 처리하고 의미를 파악하는 작업을 수행합니다. 이 작업은 텍스트 데이터를 분석하고 의미를 파악하는 것을 포함합니다.

    자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다. 텍스트 데이터는 텍스트를 의미 있는 방식으로 변환하는 과정을 포함합니다. 텍스트 데이터를 처리하기 위해 기계 학습 알고리즘을 사용합니다. 예를 들어, 텍스트 데이터를 전처리하여 의미를 파악하는 작업을 수행합니다.

    자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다. 텍스트 데이터를 전처리하여 의미를 파악하는 작업을 수행합니다. 텍스트 데이터를 전처리하는 과정은 텍스트 데이터를 분석하고 의미를 파악하는 작업을 포함합니다.

    자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다. 텍스트 데이터를 분석하고 의미를 파악하는 작업을 수행합니다. 텍스트 데이터를 분석하는 과정은 텍스트 데이터를 전처리하여 의미를 파악하는 작업을 포함합니다.

    자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다.

- 결과 비교:

    문장이 완성되지 않은 입력에 대해 자연어 생성을 목적으로 하는 GPT 모델은 문장이 완성되도록 입력의 뒤에 창의적이고 다양한 내용을 생성합니다. 챗봇 용도의 GPT 모델은 사용자에게 도움이 되고 유익한 내용을 제공하라는 프롬프트에 맞게 단지 입력에 나와있는 자연어 처리에 대한 내용을 설명합니다.

In [ ]:
# (자연어 생성에 적합할 것으로 기대되는) 입력 문장1
sent = "자연어 처리에서 최근" ### 이 부분을 새로운 문장으로 작성

In [ ]:
# 자연어 생성 (sent) 출력
pipe = pipeline("text-generation", model="skt/kogpt2-base-v2", device=0)
generated_text = pipe(sent, max_new_tokens=256)
print(generated_text[0]['generated_text'])

자연어 처리에서 최근 각광받고 있는 ‘디지틀 프로그래밍 언어’라며 컴퓨터 언어의 중요성을 강조했다.
이에 따라 이 책은 컴퓨터를 제대로 활용할 수 있는 컴퓨터 교육 방안을 제시하고 있다.
또 컴퓨터와 컴퓨터를 통해 학습할 수 없는 다양한 문제에 대한 효과적인 해결책을 제시해 컴퓨터 교육을 효과적으로 도와준다.
특히 지난해 출간된 ‘컴퓨터와 세상을 읽는 방법’ 시리즈와 2004년 초판 등에 이어 이번 학기에는 컴퓨터 교육과 관련된 서적들을 새롭게 포함시킨 ‘컴퓨터 세상을 읽는 방법’편을 새롭게 선보이는 등 전자책을 통한 독서의 중요성을 강조하고 있다.
책자에는 ‘디지틀 프로그래밍 언어’에 대한 내용뿐 아니라 일상생활에서 접하는 컴퓨터 교육과 관련된 다양한 정보를 싣는다.
이 밖에도 컴퓨터 교육을 올바르게 활용할 수 있도록 초등학교 5학년과 고등학교 2학년 과정을 위한 컴퓨터 교육과정과, 컴퓨터 교육을 위해 노력하는 사람들을 위한 컴퓨터 교육, 생활 속에서 자연스럽게 배울 수 있는 컴퓨터 사용법 소개, 컴퓨터 교육 방법 등이 자세하게 소개돼 있다. 이달부터 한국항공우주산업이 발사 후 2주 만에 궤도 진입에 성공하면서 본격적인 우주개발에 박차를 가한다.
한국항공우주산업(KAI)이 한국형 발사체인 ‘KA-3’ 제작 사업에 나선다.
한국항공우주산업은 13일 서울 코엑스 인터컨티넨탈 호텔에서 기자 간담회를 통해 위성체 개발 등 우주개발 사업을 본격화한다고 밝혔다.
이번 사업으로 KAI


In [ ]:
# 챗봇 (chat_sent) 출력
pipe = pipeline("text-generation", model="heegyu/polyglot-ko-1.3b-chat", device=0)
prompt = '당신은 AI 챗봇입니다. 사용자에게 도움이 되고 유익한 내용을 제공해야 합니다. 답변은 길고 자세하며 친절한 설명을 덧붙여서 작성하세요.'
generated_text = pipe(f"{prompt}\n\n### 사용자: {sent}\n\n### 챗봇:", max_new_tokens=256, eos_token_id=2, pad_token_id=2)
print(generated_text[0]['generated_text'].split('### 챗봇:')[1])


자연어 처리에서 최근에는 텍스트 데이터를 처리하고 의미를 파악하는 작업을 수행합니다. 이 작업은 텍스트 데이터를 분석하고 의미를 파악하는 것을 포함합니다.

자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다. 텍스트 데이터는 텍스트를 의미 있는 방식으로 변환하는 과정을 포함합니다. 텍스트 데이터를 처리하기 위해 기계 학습 알고리즘을 사용합니다. 예를 들어, 텍스트 데이터를 전처리하여 의미를 파악하는 작업을 수행합니다.

자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다. 텍스트 데이터를 전처리하여 의미를 파악하는 작업을 수행합니다. 텍스트 데이터를 전처리하는 과정은 텍스트 데이터를 분석하고 의미를 파악하는 작업을 포함합니다.

자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다. 텍스트 데이터를 분석하고 의미를 파악하는 작업을 수행합니다. 텍스트 데이터를 분석하는 과정은 텍스트 데이터를 전처리하여 의미를 파악하는 작업을 포함합니다.

자연어 처리는 텍스트 데이터를 처리하고 의미를 파악하는 과정을 포함합니다.


[테스트 입력 문장2]
- (**챗봇** 입력으로 적합할 것으로 기대되는) 입력 문장2 : "자연어 처리에서 최근 어떤 것이 주목받고 있어?"

- **자연어 생성** 출력 결과:

    자연어 처리에서 최근 어떤 것이 주목받고 있어?"라면서 "과거에는 어뢰에 의존하여 항해할 수 있는 길이 있었지만 지금은 어뢰를 이용하여 항해할 수 있는 길이 열렸다"라고 설명했다. 지난해 9월 16일부터 내달 12일까지 서울 서초동에서 열리는 이 행사는 ‘서울가족의 해’라는 주제로 마련됐다.
    올해도 어르신 400명을 대상으로 가족사진과 동영상 등 관련 교육을 진행하는 행사다.
    행사장에서는 시니어 요리사의 시니어 요리교실도 이뤄진다.
    동영상 공모에는 서울가족이 참여할 수 있다.
    서울가족이 직접 진행하는 요리경연에 신청하면 된다.
    본선대회는 다음 달 27일 열린다.
    서울가족은 시제품 심사를 위해 다양한 조리방법을 시도한다.
    우선 어르신 요리경연은 ‘혼밥이 좋다’로 가장 먼저 치러진다.
    둘째 날은 ‘친구들과 함께하는 요리’를 주제로 두 차례 진행한다.
    둘째 날은 ‘보이는 식탁 요리’를 주제로, 셋째 날은 ‘고급스러운 인테리어 만들기’를 주제로 열띤 경쟁을 펼치게 된다.
    요리사 2급 이상이면 누구나 참가할 수 있으며 참가비는 1인 3만원이고, 가족사진촬영까지 할 수 있다.
    참가자 중 추첨을 통해 200명에게 다양한 경품을 제공한다.
    시작에 앞서 행사에서는 간단한 퀴즈를 풀고 어르신들의 시선으로 요리과정을 지켜본다.
    이후 참가자에게는 사진촬영과

- **챗봇** 출력 결과:

    최근 자연어 처리에서 주목받고 있는 몇 가지 기술은 다음과 같습니다.

    1. 딥러닝: 딥러닝은 기계 학습 알고리즘을 사용하여 데이터를 학습하고 예측하는 기술입니다. 이는 텍스트 데이터를 처리하는 데 사용되며, 텍스트 데이터의 패턴을 인식하고 분류하는 데에 사용됩니다.

    2. 자연어 생성: 자연어 생성은 텍스트 데이터를 생성하는 기술입니다. 이는 텍스트 데이터를 생성하고, 이를 분석하여 의미를 파악하는 데 사용됩니다.

    3. 자연어 처리: 자연어 처리는 텍스트 데이터를 분석하고, 의미를 파악하는 기술입니다. 이는 텍스트 데이터를 분석하고, 의미를 파악하는 데 사용됩니다.

    4. 기계 학습: 기계 학습은 데이터를 학습하고 예측하는 기술입니다. 이는 데이터를 학습하고, 예측하는 데 사용됩니다.

    5. 텍스트 분류: 텍스트 분류는 텍스트 데이터를 분류하는 기술입니다. 이는 텍스트 데이터를 분류하는 데 사용됩니다.

    이러한 기술들은 자연어 처리 분야에서 중요한 역할을 수행하고 있으며, 앞으로도 계속해서 발전할 것으로 예상됩니다.

- 결과 비교:

    대화 형식으로 된 입력에 대해 답변이 출력되어야 하지만 자연어 생성을 목적으로 하는 GPT 모델은 입력을 이어받아 무작위적이고 비맥락적인 내용을 생성했습니다. 반면에 챗봇 용도의 GPT 모델은 입력된 질문에 대해 주제에 맞는 적절한 답변을 생성했으며 답변은 길고 자세하며 친절한 설명을 덧붙여서 작성하라는 프롬프트에 맞게 자연어 처리에서 주목받는 기술들을 구체적으로 나열하고 각 기술에 대한 간단한 설명을 제공했습니다.

In [ ]:
# (챗봇 입력으로 적합할 것으로 기대되는) 입력 문장2
chat_sent = "자연어 처리에서 최근 어떤 것이 주목받고 있어?" ### 이 부분을 새로운 문장으로 작성

In [ ]:
# 자연어 생성 출력
pipe = pipeline("text-generation", model="skt/kogpt2-base-v2", device=0)
generated_text = pipe(chat_sent, max_new_tokens=256)
print(generated_text[0]['generated_text'])

자연어 처리에서 최근 어떤 것이 주목받고 있어?"라면서 "과거에는 어뢰에 의존하여 항해할 수 있는 길이 있었지만 지금은 어뢰를 이용하여 항해할 수 있는 길이 열렸다"라고 설명했다. 지난해 9월 16일부터 내달 12일까지 서울 서초동에서 열리는 이 행사는 ‘서울가족의 해’라는 주제로 마련됐다.
올해도 어르신 400명을 대상으로 가족사진과 동영상 등 관련 교육을 진행하는 행사다.
행사장에서는 시니어 요리사의 시니어 요리교실도 이뤄진다.
동영상 공모에는 서울가족이 참여할 수 있다.
서울가족이 직접 진행하는 요리경연에 신청하면 된다.
본선대회는 다음 달 27일 열린다.
서울가족은 시제품 심사를 위해 다양한 조리방법을 시도한다.
우선 어르신 요리경연은 ‘혼밥이 좋다’로 가장 먼저 치러진다.
둘째 날은 ‘친구들과 함께하는 요리’를 주제로 두 차례 진행한다.
둘째 날은 ‘보이는 식탁 요리’를 주제로, 셋째 날은 ‘고급스러운 인테리어 만들기’를 주제로 열띤 경쟁을 펼치게 된다.
요리사 2급 이상이면 누구나 참가할 수 있으며 참가비는 1인 3만원이고, 가족사진촬영까지 할 수 있다.
참가자 중 추첨을 통해 200명에게 다양한 경품을 제공한다.
시작에 앞서 행사에서는 간단한 퀴즈를 풀고 어르신들의 시선으로 요리과정을 지켜본다.
이후 참가자에게는 사진촬영과


In [ ]:
# 챗봇 출력
pipe = pipeline("text-generation", model="heegyu/polyglot-ko-1.3b-chat", device=0)
prompt = '당신은 AI 챗봇입니다. 사용자에게 도움이 되고 유익한 내용을 제공해야 합니다. 답변은 길고 자세하며 친절한 설명을 덧붙여서 작성하세요.'
generated_text = pipe(f"{prompt}\n\n### 사용자: {chat_sent}\n\n### 챗봇:", max_new_tokens=256, eos_token_id=2, pad_token_id=2)
print(generated_text[0]['generated_text'].split('### 챗봇:')[1])

 최근 자연어 처리에서 주목받고 있는 몇 가지 기술은 다음과 같습니다.

1. 딥러닝: 딥러닝은 기계 학습 알고리즘을 사용하여 데이터를 학습하고 예측하는 기술입니다. 이는 텍스트 데이터를 처리하는 데 사용되며, 텍스트 데이터의 패턴을 인식하고 분류하는 데에 사용됩니다.

2. 자연어 생성: 자연어 생성은 텍스트 데이터를 생성하는 기술입니다. 이는 텍스트 데이터를 생성하고, 이를 분석하여 의미를 파악하는 데 사용됩니다.

3. 자연어 처리: 자연어 처리는 텍스트 데이터를 분석하고, 의미를 파악하는 기술입니다. 이는 텍스트 데이터를 분석하고, 의미를 파악하는 데 사용됩니다.

4. 기계 학습: 기계 학습은 데이터를 학습하고 예측하는 기술입니다. 이는 데이터를 학습하고, 예측하는 데 사용됩니다.

5. 텍스트 분류: 텍스트 분류는 텍스트 데이터를 분류하는 기술입니다. 이는 텍스트 데이터를 분류하는 데 사용됩니다.

이러한 기술들은 자연어 처리 분야에서 중요한 역할을 수행하고 있으며, 앞으로도 계속해서 발전할 것으로 예상됩니다.


---

# [파일] -> [다운로드] -> [.ipynb 다운로드]